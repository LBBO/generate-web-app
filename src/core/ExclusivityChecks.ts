import type { Extension } from './Extension'

export type ValidExclusivitiesResponse = {
  isValidConfiguration: true
}
export type InvalidExclusivitiesResponse = {
  isValidConfiguration: false
  errorMessages: string[]
}

export const checkExclusivities = (
  chosenExtensions: Array<Extension>,
): ValidExclusivitiesResponse | InvalidExclusivitiesResponse => {
  const errorMessages = chosenExtensions.reduce((errorMessages, extension) => {
    const forbiddenButYetChosenExtensionNames = extension.exclusiveTo
      ?.filter((otherExtension) => chosenExtensions.includes(otherExtension))
      .map((extension) => extension.name)

    if (forbiddenButYetChosenExtensionNames?.length) {
      return [
        ...errorMessages,
        `${
          extension.name
        } cannot be used along with the following extensions: ${forbiddenButYetChosenExtensionNames.join(
          ', ',
        )}`,
      ]
    } else {
      return errorMessages
    }
  }, [] as string[])

  if (errorMessages.length) {
    return {
      isValidConfiguration: false,
      errorMessages,
    }
  } else {
    return { isValidConfiguration: true }
  }
}
