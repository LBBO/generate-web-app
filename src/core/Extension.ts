import type { Observable, Subject } from 'rxjs'
import type { Answers, DistinctQuestion } from 'inquirer'
import type { ProjectMetaData } from './userDialog/PerformUserDialog'
import type { Command } from 'commander'
import type { OverrideProperties } from '../types/UtilityTypes'

export enum ExtensionCategory {
  ONLY_FOR_TESTING,
  JAVASCRIPT_FLAVOR,
  FRONTEND_FRAMEWORK,
  CSS_PREPROCESSOR,
  JAVASCRIPT_LIBRARY,
  LINTER_OR_FORMATTER,
}

export type AdditionalInformationForExtensions = {
  projectMetadata: ProjectMetaData
  chosenExtensions: Array<Extension>
}

export type Extension = {
  name: string
  description: string
  linkToDocumentation: URL
  category: ExtensionCategory

  dependsOn?: Array<Extension>
  exclusiveTo?: Array<Extension>

  promptOptions?: (
    prompts$: Subject<DistinctQuestion>,
    answers$: Observable<Answers>,
    cliOptions: Record<string, unknown>,
  ) => Observable<Record<string, unknown>> | Promise<Record<string, unknown>>

  declareCliOptions?: (program: Command) => void

  options?: Record<string, unknown>

  canBeSkipped?: (
    options: Record<string, unknown> | undefined,
    otherInformation: AdditionalInformationForExtensions,
  ) => boolean
  run: (
    options: Record<string, unknown> | undefined,
    otherInformation: AdditionalInformationForExtensions,
  ) => Promise<void>

  printUsefulInformation?: (
    options: Record<string, unknown> | undefined,
    otherInformation: AdditionalInformationForExtensions,
  ) => void
}

export type ExtensionWithSpecificOptions<
  Options extends Record<string, unknown>,
> = OverrideProperties<Extension, { options: Options }>
