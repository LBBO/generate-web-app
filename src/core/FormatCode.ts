import * as ts from 'typescript'
import type { Options as PrettierOptions } from 'prettier'
import { format } from 'prettier'
import { diffWordsWithSpace } from 'diff'

export const defaultTypescriptCompilerOptions: ts.CompilerOptions = {
  target: ts.ScriptTarget.ESNext,
  jsx: ts.JsxEmit.Preserve,
}
export const defaultTypescriptTranspileOptions: ts.TranspileOptions = {
  compilerOptions: defaultTypescriptCompilerOptions,
}

export const defaultPrettierOptions: PrettierOptions = {
  semi: false,
  singleQuote: true,
  trailingComma: 'all',
}

export const generatePrettierOptions = (
  overrides: Partial<PrettierOptions> = {},
): PrettierOptions => ({
  ...defaultPrettierOptions,
  ...overrides,
})

export const formatWithPrettier = (
  fileContent: string,
  filePath: string,
  overrides: Partial<PrettierOptions> = {},
): string => {
  return format(
    fileContent,
    generatePrettierOptions({
      ...overrides,
      filepath: filePath,
    }),
  )
}

export const convertTypeScriptToFormattedJavaScript = (
  typeScriptCode: string,
  filePath: string,
  prettierOverrides: Partial<PrettierOptions> = {},
): string => {
  const transpiledWithoutNewlines = formatWithPrettier(
    ts.transpile(typeScriptCode, defaultTypescriptCompilerOptions),
    filePath,
    prettierOverrides,
  )

  // Re-introduce whitespaces (especially newlines) that were removed by TS
  const diff = diffWordsWithSpace(typeScriptCode, transpiledWithoutNewlines)

  const withWhiteSpaces = diff.reduce((acc, part) => {
    // Removed parts should stay removed, unless they are just white space.
    // If part is just white space, let prettier take care of it afterwards
    if (part.removed) {
      // Finds all whitespace that was removed from either end of the
      // removed part. If the entire removed part was whitespace,
      // that whitespace is also in one of the two match groups.
      const match = /^(\s*)[\s\S]*?(\s*)$/g.exec(part.value)

      const whitespaceRemovedFromEndsOfPart =
        (match?.[1] ?? '') + (match?.[2] ?? '')
      return acc + whitespaceRemovedFromEndsOfPart
    } else {
      return acc + part.value
    }
  }, '')

  return formatWithPrettier(withWhiteSpaces, filePath, prettierOverrides)
}
