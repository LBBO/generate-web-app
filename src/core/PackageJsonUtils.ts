import path from 'path'
import fs from 'fs/promises'

export const getPackageJson = async (
  rootDir: string,
): Promise<Record<string, unknown>> => {
  const filePath = path.join(rootDir, 'package.json')
  const packageJsonContent = await fs.readFile(filePath)
  return JSON.parse(packageJsonContent.toString())
}

export const getBlock = async (
  rootDir: string,
  blockName: string,
): Promise<unknown> => {
  const packageJsonObject = await getPackageJson(rootDir)
  return packageJsonObject[blockName]
}

export const removeBlock = async (
  rootDir: string,
  blockName: string,
): Promise<void> => {
  const packageJsonObject = await getPackageJson(rootDir)
  const remainingContent = {
    ...packageJsonObject,
    [blockName]: undefined,
  }
  await fs.writeFile(
    path.join(rootDir, 'package.json'),
    JSON.stringify(remainingContent, null, 2) + '\n',
  )
}
