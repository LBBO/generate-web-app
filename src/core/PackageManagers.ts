import { execSync } from 'child_process'

export enum PackageManager {
  NPM = 'npm',
  YARN = 'yarn',
}

export const isNpmInstalled = (): boolean => {
  try {
    execSync('npm -v', { stdio: 'ignore' })
    return true
  } catch (e) {
    return false
  }
}

export const isYarnInstalled = (): boolean => {
  try {
    execSync('yarn -v', { stdio: 'ignore' })
    return true
  } catch (e) {
    return false
  }
}
