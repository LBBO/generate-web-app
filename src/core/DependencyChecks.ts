import { Extension } from './Extension'

export type ValidDependenciesResponse = {
  isValidConfiguration: true
}
export type InvalidDependenciesResponse = {
  isValidConfiguration: false
  errorMessages: string[]
}

export const checkDependencies = (
  chosenExtensions: Array<Extension>,
): ValidDependenciesResponse | InvalidDependenciesResponse => {
  const errorMessages = chosenExtensions.reduce((errorMessages, extension) => {
    // Check every chosen extension. If there are unmet dependencies, add a message to the array of error messages
    const missingDependencies = extension.dependsOn
      ?.filter((dependency) => !chosenExtensions.includes(dependency))
      .map((dependency) => dependency.name)

    if (missingDependencies?.length) {
      const missingDependencyNames = missingDependencies.join(', ')

      return [
        ...errorMessages,
        `${extension.name} has unmet dependencies: ${missingDependencyNames}`,
      ]
    } else {
      return errorMessages
    }
  }, [] as string[])

  if (errorMessages.length) {
    return {
      isValidConfiguration: false,
      errorMessages,
    }
  } else {
    return { isValidConfiguration: true }
  }
}
