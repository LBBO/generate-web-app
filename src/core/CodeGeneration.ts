import { readFile, writeFile } from 'fs/promises'
import { formatWithPrettier } from './FormatCode'

export type JsOrTsImportData = {
  importDefault?: string
  importItems?: string[]
  importAllAs?: string
  sourcePath: string
}

export const addImportToJsOrTsFile = async (
  filePath: string,
  importData: JsOrTsImportData,
): Promise<void> => {
  const fileContent = (await readFile(filePath)).toString()
  const fileContentWithSingleLineInputs = formatWithPrettier(
    fileContent,
    filePath,
    {
      // Forces multiline-inputs to be single-line
      printWidth: Infinity,
    },
  )
  const initialImportsRegex = /^(?:import\W.*\n|\s)*/g
  const initialImports = initialImportsRegex
    .exec(fileContentWithSingleLineInputs)?.[0]
    ?.trim()

  if (initialImports === undefined) {
    throw new Error(`Couldn't find any imports at beginning of line`)
  } else {
    let valuesToImport: string | undefined

    if (importData.importAllAs !== undefined) {
      valuesToImport = `* as ${importData.importAllAs}`
    } else {
      const partsOfValuesToImport: string[] = []

      if (importData.importDefault) {
        partsOfValuesToImport.push(importData.importDefault)
      }

      if (importData.importItems) {
        partsOfValuesToImport.push(
          '{ ' +
            importData.importItems
              .sort((a, b) => {
                if (a < b) {
                  return -1
                } else if (a > b) {
                  return 1
                } else {
                  return 0
                }
              })
              .join(', ') +
            ' }',
        )
      }

      valuesToImport = partsOfValuesToImport.join(', ')
    }

    const newImport = `
import ${valuesToImport ? valuesToImport + ' from' : ''} '${
      importData.sourcePath
    }'
`
    const newFileContent =
      initialImports +
      newImport +
      fileContentWithSingleLineInputs.slice(initialImports.length + 1)

    await writeFile(filePath, formatWithPrettier(newFileContent, filePath))
  }
}
