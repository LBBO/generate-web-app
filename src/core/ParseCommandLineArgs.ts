import type { Extension } from './Extension'
import type { Command, OptionValues } from 'commander'
import type { ProjectMetaData } from './userDialog/PerformUserDialog'
import type { PackageManagerStrategy } from './packageManagers/PackageManagerStrategy'
import { PackageManagerNames } from './packageManagers/PackageManagerStrategy'
import path from 'path'
import { generateYarnPackageManagerStrategy } from './packageManagers/YarnPackageManagerStrategy'
import { generateNpmPackageManagerStrategy } from './packageManagers/NpmPackageManagerStrategy'
import { checkExclusivities } from './ExclusivityChecks'
import { checkDependencies } from './DependencyChecks'
import chalk from 'chalk'

export const declareArgsAndOptions = (
  program: Command,
  allExtensions: Array<Extension>,
): void => {
  program
    .argument(
      '[app-name]',
      'The name or path to the folder that should hold the app. The folder should not exist' +
        ' yet and will be created by generate-web-app.',
    )
    .option(
      '-p, --package-manager <package-manager-name>',
      'Choose your package manager (npm or yarn)',
    )

  allExtensions.forEach((extension) => {
    const normalDescription = `Install ${extension.name}`
    const dependencyDescription = extension.dependsOn?.length
      ? `Requires ${extension.dependsOn
          ?.map((dependency) => dependency.name)
          .join(', ')}`
      : undefined
    const exclusivityDescription = extension.exclusiveTo?.length
      ? `Cannot be used along with ${extension.exclusiveTo
          ?.map((exclusivity) => exclusivity.name)
          .join(', ')}`
      : undefined

    const finalDescription = [
      normalDescription,
      dependencyDescription,
      exclusivityDescription,
    ]
      .filter((description) => description !== undefined)
      .join(' - ')

    program.option(`--${extension.name.toLowerCase()}`, finalDescription)
    extension.declareCliOptions?.(program)
  })
}

export const parseMetaData = (
  args: string[],
  options: OptionValues,
): Partial<ProjectMetaData> => {
  const rootDirectory = [undefined, ''].includes(args[0])
    ? undefined
    : path.join(process.cwd(), args[0])

  if (rootDirectory && path.extname(rootDirectory) !== '') {
    throw new Error(`Invalid target directory: ${rootDirectory}`)
  }

  const programName =
    rootDirectory === undefined ? undefined : path.parse(rootDirectory).name

  let packageManager: PackageManagerNames | undefined = undefined
  let packageManagerStrategy: PackageManagerStrategy | undefined = undefined

  switch (options.packageManager) {
    case undefined:
      // No package manager was specified, just keep undefined
      break
    case PackageManagerNames.YARN:
      packageManager = PackageManagerNames.YARN
      if (rootDirectory) {
        packageManagerStrategy =
          generateYarnPackageManagerStrategy(rootDirectory)
      }
      break
    case PackageManagerNames.NPM:
      packageManager = PackageManagerNames.NPM
      if (rootDirectory) {
        packageManagerStrategy =
          generateNpmPackageManagerStrategy(rootDirectory)
      }
      break
    default:
      throw new Error(
        `${options.packageManager} is not a known package manager.`,
      )
  }

  return {
    rootDirectory,
    name: programName,
    chosenPackageManager: packageManager,
    packageManagerStrategy,
  }
}

export const parseChosenExtensions = (
  args: string[],
  options: OptionValues,
  allExtensions: Array<Extension>,
): Array<Extension | undefined> | undefined => {
  const chosenExtensionInOrderOfCliArg = Object.entries(options)
    // Remove options that were explicitly disabled
    .filter(([, optionValue]) => Boolean(optionValue))
    .map(([optionName]) =>
      allExtensions.find(
        (extension) => extension.name.toLowerCase() === optionName,
      ),
    )
    .filter(
      (extension: Extension | undefined): extension is Extension =>
        extension !== undefined,
    )

  const dependencyCheckResult = checkDependencies(
    chosenExtensionInOrderOfCliArg,
  )
  const exclusivityCheckResult = checkExclusivities(
    chosenExtensionInOrderOfCliArg,
  )

  if (!dependencyCheckResult.isValidConfiguration) {
    dependencyCheckResult.errorMessages.forEach((message) =>
      console.error(chalk.red(message)),
    )
  }
  if (!exclusivityCheckResult.isValidConfiguration) {
    exclusivityCheckResult.errorMessages.forEach((message) =>
      console.error(chalk.red(message)),
    )
  }

  if (
    !dependencyCheckResult.isValidConfiguration ||
    !exclusivityCheckResult.isValidConfiguration
  ) {
    throw new Error(
      'Your chosen configuration is not valid. For more info, please see previous error messages.',
    )
  }

  const chosenExtensionsInCorrectOrder = allExtensions.map((extension) =>
    chosenExtensionInOrderOfCliArg.includes(extension) ? extension : undefined,
  )

  return chosenExtensionsInCorrectOrder.filter(
    (extension) => extension !== undefined,
  ).length
    ? chosenExtensionsInCorrectOrder
    : undefined
}

export const parseCommandLineArgs = (
  program: Command,
  allExtensions: Array<Extension>,
): {
  metaData: Partial<ProjectMetaData>
  chosenExtensions: Array<Extension | undefined> | undefined
} => {
  declareArgsAndOptions(program, allExtensions)

  program.parse(process.argv)

  const options = program.opts()
  const args = program.args
  const metaData = parseMetaData(args, options)
  const chosenExtensions = parseChosenExtensions(args, options, allExtensions)

  return { metaData, chosenExtensions }
}
