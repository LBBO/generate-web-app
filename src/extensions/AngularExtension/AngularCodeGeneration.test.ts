import path from 'path'
import {
  addAngularImportToModule,
  addDeclarationToModule,
} from './AngularCodeGeneration'
import type { JsOrTsImportData } from '../../core/CodeGeneration'
import * as GeneralCodeGeneration from '../../core/CodeGeneration'
import fs from 'fs/promises'
import {
  appModuleWithMoreDeclarations,
  appModuleWithMoreImports,
  defaultAppModule,
} from './AngularCodeGenerationMockData'

const defaultPathToAppModule = path.join(__dirname, 'app.module.ts')
const defaultComponentImportData: JsOrTsImportData = {
  sourcePath: './my-component/my-component.component',
  importItems: ['MyComponent'],
}
const defaultModuleImportData: JsOrTsImportData = {
  sourcePath: '@angular/forms',
  importItems: ['FormsModule'],
}

let addImportToJsOrTsFileMock: jest.SpyInstance
let readFileMock: jest.SpyInstance
let writeFileMock: jest.SpyInstance

beforeEach(() => {
  addImportToJsOrTsFileMock = jest
    .spyOn(GeneralCodeGeneration, 'addImportToJsOrTsFile')
    .mockResolvedValue(undefined)
  readFileMock = jest.spyOn(fs, 'readFile').mockResolvedValue(defaultAppModule)
  writeFileMock = jest.spyOn(fs, 'writeFile').mockResolvedValue()
})

describe('addDeclarationToModule', () => {
  it('should create an item import for the declaration if desired', async () => {
    await addDeclarationToModule(
      defaultPathToAppModule,
      defaultComponentImportData,
    )

    expect(addImportToJsOrTsFileMock).toHaveBeenCalledTimes(1)
    expect(addImportToJsOrTsFileMock).toHaveBeenCalledWith(
      defaultPathToAppModule,
      defaultComponentImportData,
    )
  })

  it('should correctly add declarations to the AppModule', async () => {
    await addDeclarationToModule(
      defaultPathToAppModule,
      defaultComponentImportData,
    )

    expect(writeFileMock).toHaveBeenCalledTimes(1)
    expect(writeFileMock.mock.calls[0][1]).toMatchSnapshot()
  })

  it('should sort declarations alphabetically', async () => {
    readFileMock.mockResolvedValue(appModuleWithMoreDeclarations)
    await addDeclarationToModule(
      defaultPathToAppModule,
      defaultComponentImportData,
    )

    expect(writeFileMock.mock.calls[0][1]).toMatchSnapshot()
  })
})

describe('addAngularImportToModule', () => {
  it('should create an item import for the import if desired', async () => {
    await addAngularImportToModule(
      defaultPathToAppModule,
      defaultModuleImportData,
    )

    expect(addImportToJsOrTsFileMock).toHaveBeenCalledTimes(1)
    expect(addImportToJsOrTsFileMock.mock.calls[0][1]).toBe(
      defaultModuleImportData,
    )
  })

  it('should correctly add imports to the AppModule', async () => {
    await addAngularImportToModule(
      defaultPathToAppModule,
      defaultModuleImportData,
    )

    expect(writeFileMock).toHaveBeenCalledTimes(1)
    expect(writeFileMock.mock.calls[0][1]).toMatchSnapshot()
  })

  it('should sort imports alphabetically', async () => {
    readFileMock.mockResolvedValue(appModuleWithMoreImports)

    await addAngularImportToModule(
      defaultPathToAppModule,
      defaultModuleImportData,
    )

    expect(writeFileMock.mock.calls[0][1]).toMatchSnapshot()
  })
})

describe('addAngularComponent', () => {
  it.todo('should add the component in the correct position')
})
