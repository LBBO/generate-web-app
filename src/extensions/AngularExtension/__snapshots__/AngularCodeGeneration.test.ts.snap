// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addAngularImportToModule should correctly add imports to the AppModule 1`] = `
"import { NgModule } from '@angular/core'
import { BrowserModule } from '@angular/platform-browser'

import { AppComponent } from './app.component'

@NgModule({
  declarations: [AppComponent],
  imports: [BrowserModule, FormsModule],
  providers: [],
  bootstrap: [AppComponent],
})
export class AppModule {}
"
`;

exports[`addAngularImportToModule should sort imports alphabetically 1`] = `
"import { NgModule } from '@angular/core'
import { BrowserModule } from '@angular/platform-browser'

import { AppComponent } from './app.component'

@NgModule({
  declarations: [AppComponent],
  imports: [
    AModule,
    AnotherModule,
    BrowserModule,
    CoolModule,
    DemonstrationModule,
    FormsModule,
    TestingModule,
    YoMamasModule,
    ZebraModule,
  ],
  providers: [],
  bootstrap: [AppComponent],
})
export class AppModule {}
"
`;

exports[`addDeclarationToModule should correctly add declarations to the AppModule 1`] = `
"import { NgModule } from '@angular/core'
import { BrowserModule } from '@angular/platform-browser'

import { AppComponent } from './app.component'

@NgModule({
  declarations: [AppComponent, MyComponent],
  imports: [BrowserModule],
  providers: [],
  bootstrap: [AppComponent],
})
export class AppModule {}
"
`;

exports[`addDeclarationToModule should sort declarations alphabetically 1`] = `
"import { NgModule } from '@angular/core'
import { BrowserModule } from '@angular/platform-browser'

import { AppComponent } from './app.component'

@NgModule({
  declarations: [
    AnotherComponent,
    AppComponent,
    CreativelyNamedComponent,
    MagicalComponent,
    MyComponent,
    SomeComponent,
  ],
  imports: [BrowserModule],
  providers: [],
  bootstrap: [AppComponent],
})
export class AppModule {}
"
`;
